code/controllers/subsystem/dbcore.dm

```diff
@@ -0,0 +1,364 @@
+		to_chat(world, "<span class='boldannounce'>Database schema ([sql_version]) doesn't match the latest schema version ([SQL_VERSION]). Roundstart has been delayed.</span>")
+
+	return ..()
+
+/datum/controller/subsystem/dbcore/fire()
+	for(var/I in active_queries)
+		var/datum/db_query/Q = I
+		if(world.time - Q.last_activity_time > (5 MINUTES))
+			message_admins("Found undeleted query, please check the server logs and notify coders.")
+			log_sql("Undeleted query: \"[Q.sql]\" LA: [Q.last_activity] LAT: [Q.last_activity_time]")
+			qdel(Q)
+		if(MC_TICK_CHECK)
+			return
+
+/datum/controller/subsystem/dbcore/Recover()
+	connection = SSdbcore.connection
+
+//nu
+/datum/controller/subsystem/dbcore/can_vv_get(var_name)
+	return var_name != NAMEOF(src, connection) && var_name != NAMEOF(src, active_queries) && ..()
+
+/datum/controller/subsystem/dbcore/vv_edit_var(var_name, var_value)
+	if(var_name == NAMEOF(src, connection))
+		return FALSE
+	return ..()
+
+/datum/controller/subsystem/dbcore/proc/Connect()
+	if(IsConnected())
+		return TRUE
+
+	if(failed_connection_timeout <= world.time) //it's been more than 5 seconds since we failed to connect, reset the counter
+		failed_connections = 0
+
+	if(failed_connections > 5)	//If it failed to establish a connection more than 5 times in a row, don't bother attempting to connect for 5 seconds.
+		failed_connection_timeout = world.time + 50
+		return FALSE
+
+	if(!config.sql_enabled)
+		return FALSE
+
+	var/user = sqlfdbklogin
+	var/pass = sqlfdbkpass
+	var/db = sqlfdbkdb
+	var/address = sqladdress
+	var/port = text2num(sqlport)
+	var/timeout = config.async_sql_query_timeout
+	var/thread_limit = config.rust_sql_thread_limit
+
+	var/result = json_decode(rustg_sql_connect_pool(json_encode(list(
+		"host" = address,
+		"port" = port,
+		"user" = user,
+		"pass" = pass,
+		"db_name" = db,
+		"read_timeout" = timeout,
+		"write_timeout" = timeout,
+		"max_threads" = thread_limit,
+	))))
+	. = (result["status"] == "ok")
+	if (.)
+		connection = result["handle"]
+	else
+		connection = null
+		last_error = result["data"]
+		log_sql("Connect() failed | [last_error]")
+		++failed_connections
+
+/datum/controller/subsystem/dbcore/proc/CheckSchemaVersion()
+	if(config.sql_enabled)
+		// The unit tests have their own version of this check, which wont hold the server up infinitely, so this is disabled if we are running unit tests
+		#ifndef UNIT_TESTS
+		if(config.sql_enabled && sql_version != SQL_VERSION)
+			config.sql_enabled = FALSE
+			schema_valid = FALSE
+			SSticker.ticker_going = FALSE
+			log_world("Database connection failed: Invalid SQL Versions")
+			return FALSE
+		#endif
+		if(Connect())
+			log_world("Database connection established")
+		else
+			// log_sql() because then an error will be logged in the same place
+			log_sql("Your server failed to establish a connection with the database")
+	else
+		log_sql("Database is not enabled in configuration")
+
+/datum/controller/subsystem/dbcore/proc/Disconnect()
```

Comment:
Nope, but this is called at the very end of the game lifetime. 
![image](https://user-images.githubusercontent.com/25063394/100525670-83368b00-31ba-11eb-826b-5441e2f0d937.png)

If there are any queries at this point, they likely wouldnt have enough time to finish anyway. That and no crucial queries run at this point anyway. Blackbox saving happens before this, and even though the queries are async as far as DD is concerned, it will still halt the executing proc in place until the query has finished.